test/api/apikeys_test.go

```diff
@@ -0,0 +1,127 @@
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/SkynetLabs/skynet-accounts/database"
+	"github.com/SkynetLabs/skynet-accounts/test"
+)
+
+// testAPIKeysFlow validates the creation, listing, and deletion of API keys.
+func testAPIKeysFlow(t *testing.T, at *test.AccountsTester) {
+	name := test.DBNameForTest(t.Name())
+	r, _, err := at.CreateUserPost(name+"@siasky.net", name+"_pass")
+	if err != nil {
+		t.Fatal(err)
+	}
+	at.Cookie = test.ExtractCookie(r)
+
+	aks := make([]database.APIKey, 0)
+
+	// List all API keys this user has. Expect the list to be empty.
+	r, body, err := at.Get("/user/apikey", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = json.Unmarshal(body, &aks)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(aks) > 0 {
+		t.Fatalf("Expected an empty list of API keys, got %+v.", aks)
+	}
+
+	// Create a new API key.
+	r, body, err = at.Post("/user/apikey", nil, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var ak1 database.APIKey
+	err = json.Unmarshal(body, &ak1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create another API key.
+	r, body, err = at.Post("/user/apikey", nil, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var ak2 database.APIKey
+	err = json.Unmarshal(body, &ak2)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// List all API keys this user has. Expect to find both keys we created.
+	r, body, err = at.Get("/user/apikey", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = json.Unmarshal(body, &aks)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(aks) != 2 {
+		t.Fatalf("Expected two API keys, got %+v.", aks)
+	}
+	if ak1.Key != aks[0].Key && ak1.Key != aks[1].Key {
+		t.Fatalf("Missing key '%s'! Set: %+v", ak1.Key, aks)
+	}
+	if ak2.Key != aks[0].Key && ak2.Key != aks[1].Key {
+		t.Fatalf("Missing key '%s'! Set: %+v", ak2.Key, aks)
+	}
+
+	// Delete an API key.
+	r, body, err = at.Delete("/user/apikey/"+ak1.Key, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// List all API keys this user has. Expect to find only the second one.
+	r, body, err = at.Get("/user/apikey", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = json.Unmarshal(body, &aks)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(aks) != 1 {
+		t.Fatalf("Expected one API key, got %+v.", aks)
+	}
+	if ak2.Key != aks[0].Key {
+		t.Fatalf("Missing key '%s'! Set: %+v", ak2.Key, aks)
+	}
+}
+
+// testAPIKeysUsage makes sure that we can use API keys to make API calls.
+func testAPIKeysUsage(t *testing.T, at *test.AccountsTester) {
+	name := test.DBNameForTest(t.Name())
+	// Create a test user.
+	_, _, err := at.CreateUserPost(name+"@siasky.net", name+"_pass")
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Create a new API key.
+	_, body, err := at.Post("/user/apikey", nil, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var ak database.APIKey
+	err = json.Unmarshal(body, &ak)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Get user stats without a cookie or headers - pass the API key via a query
+	// variable.
+	_, body, err = at.Get("/user/stats?api_key="+ak.Key, nil)
```

Comment:
are we being consistent here? just want to make sure the `_` is well thought through

Reply:
I agree that this goes against our conventions. The reason I used this is that wherever I read about API keys, I always stumbled on the Swagger standard for those (https://swagger.io/docs/specification/authentication/api-keys/) and that works with the `X-API-KEY` header or the `api_key` query param. It's a choice between compatibility with a well-established web standard and our conventions. I'm very much open to suggestions here - I'd be happy to change that to `apiKey`, if that's the consensus.

Reply:
I wouldn't go out of our way to try and be compliant with some standard that's mostly useful for autogenerating APIs if we don't plan on doing that ourselves anyway. We probably diverge in way too many places already anyway.

Reply:
Swagger has been on my mind for quite a while now. I've used if before and it's some of the best API docs I've ever used. Combined with a developer portal with allows you to make API calls to a test server from a web page by pasting your request bodies/params, it makes for an amazing dev tool.

That said, I don't think there's anything preventing us from using Swagger in the future with non-standard headers. If you prefer the `X-SKYNET-API_KEY`/`apiKey` approach, I don't mind.